sycl/include/CL/sycl/detail/pi.h

```diff
@@ -787,6 +775,23 @@ pi_result piDevicePartition(pi_device device,
                             pi_uint32 num_devices, pi_device *out_devices,
                             pi_uint32 *out_num_devices);
 
+/// Gets the native handle of a PI device object.
+///
+/// \param device is the PI device to get the native handle of.
+/// \param nativeHandle is the native handle of device.
+pi_result piextDeviceGetNativeHandle(pi_device device,
+                                     pi_native_handle *nativeHandle);
+
+/// Creates PI device object from a native handle.
+/// NOTE: If the created PI object shares reference count with the native object
```

Comment:
How would a user tell if it is one or another? I think we should be very specific and say that we start reference-counting the created PI object, i.e. it is returned to the user after being "retained". We should further say what will happen to the underlying native object once this PI object is released. For this I see 2 possibilities: 1) nothing, i.e. PI will not do anything to the native handle and it is users' responsibility to track/destroy it, and 2) PI will force-destroy the native handle disregarding its any other possible users, i.e. assumes ownership.

Reply:
> How would a user tell if it is one or another? I think we should be very specific and say that we start reference-counting the created PI object, i.e. it is returned to the user after being "retained".

That is close to the original wording of this, but I fear that there could be confusion as to what "retaining" a new PI object without shared reference count means. Does it mean that it is created with at reference count of 1 and then the it is incremented to 2, or does it simply mean that it is created with a reference count of 1? In this case I believe it should the latter, but I can see a user drawing either conclusion.

> We should further say what will happen to the underlying native object once this PI object is released. For this I see 2 possibilities: 1) nothing, i.e. PI will not do anything to the native handle and it is users' responsibility to track/destroy it, and 2) PI will force-destroy the native handle disregarding its any other possible users, i.e. assumes ownership.

To my knowledge the current OpenCL BE does neither of these, right? It uses the reference counting of the native object, so it can die in either user code or in PI based on which uses it last. The first of your suggestions could be implemented by short-circuiting the retain and release functions of the OpenCL BE for PI objects created this way.

Personally I'm more in favor of the latter, if by ownership we mean that it takes ownership of the one reference the user holds to the object. From a user's perspective this would mean that they must keep the PI object alive for the time they want to keep using their native object, but OpenCL operations can still keep it alive longer if need be.

Reply:
I am fine with #2 as well, especially because users don't know about PI objects lifetime, meaning they can't really tell when PI is done with using their handles. Let's document/implement it this way, not sure how to achieve that in OpenCL and if this is aligned with the current/future specification.

Reply:
For the OpenCL backend I think it would be enough to just have the create-from-native functions not retain the native object and just take ownership of that reference. We can't stop the user from retaining the object themselves and if it has been retained by an OpenCL operation the user passed the native handle to it is probably better that it gets to stay alive for the duration of that operation as well.

Before I go ahead and change the implementation to reflect this I want to check with @garimagu

Reply:
I think we should get this moving, so I've added a new commit with the changes as in suggestion 2). Please have a look and I will fix the merge conflicts.

Reply:
I fail to understand how moving the retains out of OpenCL PI plugin into OpenCL interoperability constructors makes any difference. Also the OpenCL object ownership is not acquired by PI, right? The native handle will get destroyed by OpenCL RT when reference counter gets to zero.

Reply:
The idea is to avoid the need for an explicit user call to release. For OpenCL I believe this is the best we can do, however the SYCL spec requires that any OpenCL interop constructor retains the corresponding OpenCL object so with this change we need to make sure that if we have one of these interoperability constructors we make sure that PI does not take sole ownership of the OpenCL object, which we can do without breaking it since it uses a shared reference count. There is no corresponding CUDA interoperability constructor so we are fine in that regard.

This means that when we start exposing the `make` functions from the aforementioned proposal the behavior is the same for CUDA and for OpenCL, namely that as after creating a SYCL object with a native handle the user need not release it. Of course if they've retained it additional times they will still have to release it the same number of times, but we can't do much in such case (as this could be in libraries or other ongoing OpenCL calls that will try to release upon completion).

Reply:
How calling cl*Retain in OpenCL constrictors is better than calling them in the piext*CreateWithNativeHandle ? I think we should get rid of direct OpenCL usage in SYCL, so can we have these back to PI level, please?

I think we should summarize this discussion briefly in the comment to piext*CreateWithNativeHandle, create a new issue to track it, and move forward with this PR.

Reply:
> 
> 
> How calling cl_Retain in OpenCL constrictors is better than calling them in the piext_CreateWithNativeHandle ? I think we should get rid of direct OpenCL usage in SYCL, so can we have these back to PI level, please?

Though I do agree that SYCL should not depend on OpenCL, these constructors are OpenCL interoperability and so will still depend on OpenCL. The benefit of moving `cl*Retain` to the SYCL interop constructors is that the state of the native elements after `piext*CreateWithNativeHandle` is somewhat consistent (i.e. the user need not call the last release of these objects in the respective APIs). Since we want to have `make` interop functions similar to the `get_native` in the foreseeable future I believe this will make their semantics easier to comprehend.

> 
> I think we should summarize this discussion briefly in the comment to piext*CreateWithNativeHandle, create a new issue to track it, and move forward with this PR.

Good call.

Reply:
I've opened the issue https://github.com/intel/llvm/issues/1516 and linked to this discussion. I'm unsure what to summarize in the comments of the `piext*CreateWithNativeHandle` that is not deviating from implementational notes. I think implementation discussion is best left for the newly opened issue.

Reply:
> The benefit of moving cl*Retain to the SYCL interop constructors is that the state of the native elements after piext*CreateWithNativeHandle is somewhat consistent (i.e. the user need not call the last release of these objects in the respective APIs). 

Thanks for opening the issue, we shall continue the discussion there as I still am uncertain this is the right thing to do. Will approve this PR now to make forward progress and because I don't see functional problems here.